src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataPrimitiveSerializer.cs

```diff
@@ -151,21 +152,40 @@ internal static object ConvertPrimitiveValue(object value, IEdmPrimitiveTypeRefe
             }
 
             Type type = value.GetType();
-            if (primitiveType != null && primitiveType.IsDate() && TypeHelper.IsDateTime(type))
-            {
-                Date dt = (DateTime)value;
-                return dt;
-            }
 
-            if (primitiveType != null && primitiveType.IsTimeOfDay() && TypeHelper.IsTimeSpan(type))
+            if (primitiveType != null)
             {
-                TimeOfDay tod = (TimeSpan)value;
-                return tod;
+                if (primitiveType.IsDate() && TypeHelper.IsDateTime(type))
+                {
+                    Date dt = (DateTime)value;
+                    return dt;
+                }
+
+                if (primitiveType.IsTimeOfDay() && TypeHelper.IsTimeSpan(type))
+                {
+                    TimeOfDay tod = (TimeSpan)value;
+                    return tod;
+                }
+
+                if (primitiveType is EdmDecimalTypeReference decimalTypeReference && decimalTypeReference.Scale.HasValue && value is decimal decimalValue)
+                {
+                    return CalculateDecimalScale(decimalValue, decimalTypeReference.Scale.Value);
+                }
             }
 
             return ConvertUnsupportedPrimitives(value);
         }
 
+        static decimal CalculateDecimalScale(decimal value, int scale)
+        {
+            string scaleForConverting = "0.";
+            for (int i = 0; i < scale; i++)
+            {
+                scaleForConverting += "0";
+            }
+            return decimal.Parse(value.ToString(scaleForConverting, CultureInfo.InvariantCulture), CultureInfo.InvariantCulture);
```

Comment:
Since the final size of the `scaleForConverting` string is known in advance (scale + 2), I wonder whether some string allocations can be avoided here. But maybe this is premature optimization:
```c#
decimal value = 10.0040m;
int scale = 2;
// we could have threshold for scale size beyond which we allocate an array
Span<char> scaleForConverting = stackalloc char[scale + 2];
scaleForConverting[0] = '0';
scaleForConverting[1] = '.';
for (int i = 0; i < scale; i++) {
    scaleForConverting[2 + i] = '0';
}
decimal result = decimal.Parse(
    value.ToString(new String(scaleForConverting), CultureInfo.InvariantCulture),
    CultureInfo.InvariantCulture);
```

Reply:
Alternatively, would `decimal.Round` work here?

I have compared your implementation with `decimal.Round(value, scale)`. The two use different rounding strategies when rounding a midpoint value. In the case of your method, it seems to always round up (i.e. 10.0050 will round up to 10.01 and 10.0150 will round up to 10.02 when the scale is 2). `decimal.Round`  uses the "round half to even strategy" (i.e. 10.0050 will round down to 10.00 and 10.0150 will round up to 10.02).

I'm not sure what the desired rounding strategy is, or what the spec says about this. But you can use an overload of `decimal.Round` that accepts a round strategy. The following snippet is functionally equivalent to your method (I also tried it with negative numbers):
```c#
decimal.Round(value, scale, MidpointRouting.AwayFromZero)
```

Reply:
If we can replace the above code with decimal.Round(), that would definitely be my preferred approach.

The [round function](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_round), at least, is defined in OData as rounding away from zero, so it would probably be consistent to do that here as well.